src/sonic-ctrmgrd/ctrmgr/container

```diff
@@ -0,0 +1,367 @@
+#!/usr/bin/python3 -u
+
+import argparse
+import docker
+import inspect
+import json
+import os
+import sys
+import syslog
+import time
+
+from swsscommon import swsscommon
+
+sys.path.append('/usr/share/sonic/scripts/')
+import container_state
+
+state_db = None
+
+# DB field names
+FEATURE_TABLE = "FEATURE"
+SET_OWNER = "set_owner"
+NO_FALLBACK = "no_fallback_to_local"
+
+CURRENT_OWNER = "current_owner"
+UPD_TIMESTAMP = "update_time"
+DOCKER_ID = "container_id"
+REMOTE_STATE = "remote_state"
+VERSION = "version"
+SYSTEM_STATE = "system_state"
+
+KUBE_LABEL_TABLE = "KUBE_LABELS"
+KUBE_LABEL_SET_KEY = "SET"
+KUBE_LABEL_UNSET_KEY = "UNSET"
+
+# Get seconds to wait for remote docker to start.
+# If not, revert to local
+#
+SONIC_CTR_CONFIG = "/etc/sonic/remote_ctr.config.json"
+SONIC_CTR_CONFIG_PEND_SECS = "revert_to_local_on_wait_seconds"
+DEFAULT_PEND_SECS = ( 5 * 60 )
+WAIT_POLL_SECS = 2
+
+def debug_msg(m):
+    msg = "{}: {}".format(inspect.stack()[1][3], m)
+    # print(msg)
+    syslog.syslog(syslog.LOG_DEBUG, msg)
+
+
+def init():
+    """ Get DB connections """
+    global state_db, cfg_db
+
+    cfg_db = swsscommon.DBConnector("CONFIG_DB", 0)
+    state_db = swsscommon.DBConnector("STATE_DB", 0)
+
+
+def get_config_data(fld, dflt):
+    """ Read entry from kube config file """
+    if os.path.exists(SONIC_CTR_CONFIG):
+         with open(SONIC_CTR_CONFIG, "r") as s:
+            d = json.load(s)
+            if fld in d:
+                return d[fld]
+    return dflt
+
+
+def read_data(is_config, feature, fields):
+    """ Read data from DB for desired fields using given defaults"""
+    ret = []
+
+    db = cfg_db if is_config else state_db
+
+    tbl = swsscommon.Table(db, FEATURE_TABLE)
+
+    data = dict(tbl.get(feature)[1])
+    for (field, default) in fields:
+        val = data.get(field, default)
+        ret += [val]
+
+    debug_msg("config:{} feature:{} fields:{} val:{}".format(
+        is_config, feature, str(fields), str(ret)))
+
+    return tuple(ret)
+
+
+def read_config(feature):
+    """ Read requried feature config """
+    set_owner, no_fallback  = read_data(True, feature, 
+            [(SET_OWNER, "local"), (NO_FALLBACK, False)])
+
+    return (set_owner, not no_fallback)
+
+
+def read_state(feature):
+    """ Read requried feature state """
+
+    return read_data(False, feature,
+            [(CURRENT_OWNER, "none"), (REMOTE_STATE, "none"), (DOCKER_ID, "")])
+
+
+def docker_action(action, feature):
+    """ Execute docker action """
+    try:
+        client = docker.from_env()
+        container = client.containers.get(feature)
+        getattr(container, action)()
+        debug_msg("docker cmd: {} for {}".format(action, feature))
+        return 0
+
+    except (docker.errors.NotFound, docker.errors.APIError) as err:
+        syslog.syslog(syslog.LOG_ERR, "docker cmd: {} for {} failed with {}".
+                format(action, feature, str(err)))
+        return -1
+
+
+def set_label(feature, create):
```

Comment:
i am a little curious about this implementation. not sure why you have two tables KUBE_LABEL_SET_KEY and KUBER_LABLE_UNSET_KEY to implment the label. ideally, you can just have one table call KUBE_LABLE|feature and then add label and delete label there.

Reply:
Points to note (before reasining):
- Kubernetes API server only has "set" labels (of course) and act on those.
- Labels we set from the switch (as we keep it in STATE-DB) is only a subset of all labels on a node.
-- Kube master or some script at master could add label to nodes
- API server is not necessarily reachable at ***all*** times.
-- In other words add/drop labels from switch can't be synchronous *always*

Reasoning:
If we drop a label, unless we have a record, we can't carry that out at a later time point, as labels at API server on a node, can be superset of what node added.

Said that:
I see a window of small problem here. If node intends to remove a label, it could remove, but it can't block master from adding the same. So instead of removing, add that label with a different value. This would block any script at master to add the same, as you can't change the label value. This being a bit of work, I would like to take it in next PR. 

Since we can't

Reply:
I have carried out what I suggested above. Now only one key for labels. Changed code will be commited upon testing